v2/APIError_test.go

```diff
@@ -0,0 +1,281 @@
+		Subject:     "Quotafailure",
+		Description: "You have exhuasted the quota",
+	}
+	qf.Violations = append(qf.Violations, v)
+	msg := ErrDetails{}
+	st, _ := status.New(codes.ResourceExhausted, "Per user quota has been exhausted").WithDetails(qf)
+	err := st.Err()
+	msg.QuotaFailure = qf
+	d := APIError{
+		err:     err,
+		status:  st,
+		details: msg,
+	}
+	assert.Equal(t, d.details, d.Details())
+}
+
+func TestUnwrap(t *testing.T) {
+	br := &errdetails.BadRequest{}
+	desc := "The argument is invalid"
+	v := &errdetails.BadRequest_FieldViolation{
+		Field:       "username",
+		Description: desc,
+	}
+	br.FieldViolations = append(br.FieldViolations, v)
+	st, _ := status.New(codes.InvalidArgument, "Invalid argument").WithDetails(br)
+	err := st.Err()
+	msg := ErrDetails{}
+	msg.BadRequest = br
+	d := APIError{
+		err:     err,
+		status:  st,
+		details: msg,
+	}
+	assert.Equal(t, d.err, d.Unwrap())
+}
+func TestError(t *testing.T) {
+	pf := &errdetails.PreconditionFailure{}
+	viol := &errdetails.PreconditionFailure_Violation{
+		Type:    "Precondition Failure",
+		Subject: "This is a test",
+	}
+	pf.Violations = append(pf.Violations, viol)
+	mesg := ErrDetails{}
+	mesg.PreconditionFailure = pf
+	stat, _ := status.New(codes.FailedPrecondition, "System's state is not suitable for operation execution").WithDetails(pf)
+	er := stat.Err()
+	d := APIError{
+		err:     er,
+		status:  stat,
+		details: mesg,
+	}
+	//str := err.Error() + "\n" + br.String()
+	assert.Contains(t, d.Error(), "System's state is not suitable for operation execution")
+}
+
+func TestGRPCStatus(t *testing.T) {
+	qf := &errdetails.QuotaFailure{}
+	v := &errdetails.QuotaFailure_Violation{
+		Subject:     "Quotafailure",
+		Description: "You have exhuasted the quota",
+	}
+	qf.Violations = append(qf.Violations, v)
+	st, _ := status.New(codes.ResourceExhausted, "Per user quota has been exhausted").WithDetails(qf)
+	err := st.Err()
+	msg := ErrDetails{}
+	msg.QuotaFailure = qf
+	d := APIError{
+		err:     err,
+		status:  st,
+		details: msg,
+	}
+	assert.Equal(t, d.status, d.GRPCStatus())
+
+}
+func TestNilFromError(t *testing.T) {
+	err := error(nil)
+	_, ok := FromError(err)
+	if ok == true {
+		fmt.Println("Test Failed. Nil error should not be wrapped as API")
+	}
+}
+
+func TestNonAPIFromError(t *testing.T) {
+	err := context.DeadlineExceeded
+	_, ok := FromError(err)
+	if ok == true {
+		fmt.Println("Test Failed. Do not wrap nonAPI errors")
+	}
+}
+
+func TestAPIFromError(t *testing.T) {
+	type test struct {
+		name string
+		want string
+		got  string
+	}
+	br := &errdetails.BadRequest{}
+	br_violation := &errdetails.BadRequest_FieldViolation{
+		Field:       "field",
+		Description: "desc",
+	}
+	br.FieldViolations = append(br.FieldViolations, br_violation)
+	br_stat, _ := status.New(codes.InvalidArgument, "bad request").WithDetails(br)
+	br_err := br_stat.Err()
+	br_msg := ErrDetails{}
+	br_msg.BadRequest = br
+	br_expected := APIError{
```

Comment:
Please try to initialize all structs in this manner, rather than individually setting each field on a separate line. You can initialize nested structs all in one big block too.